hashprobe/hashprobe.go

```diff
@@ -0,0 +1,643 @@
+
+import (
+	"math"
+	"math/bits"
+	"math/rand"
+
+	"github.com/segmentio/parquet-go/hashprobe/aeshash"
+	"github.com/segmentio/parquet-go/hashprobe/wyhash"
+	"github.com/segmentio/parquet-go/internal/unsafecast"
+)
+
+const (
+	// Number of probes tested per iteration. This parameter balances between
+	// the amount of memory allocated on the stack to hold the computed hashes
+	// of the keys being probed, and amortizing the baseline cost of the probing
+	// algorithm.
+	//
+	// The larger the value, the more memory is required, but lower the baseline
+	// cost will be.
+	//
+	// We chose a value that is somewhat large, resulting in reserving 2KiB of
+	// stack but mostly erasing the baseline cost.
+	probesPerLoop = 256
+)
+
+func nextPowerOf2(n int) int {
+	return 1 << (64 - bits.LeadingZeros64(uint64(n-1)))
+}
+
+func randSeed() uintptr {
+	return uintptr(rand.Uint64())
+}
+
+type Int32Table struct{ table32 }
+
+func NewInt32Table(cap int, maxLoad float64) *Int32Table {
+	return &Int32Table{makeTable32(cap, maxLoad)}
+}
+
+func (t *Int32Table) Reset() { t.reset() }
+
+func (t *Int32Table) Len() int { return t.len }
+
+func (t *Int32Table) Cap() int { return t.cap }
+
+func (t *Int32Table) Probe(keys, values []int32) int {
+	return t.probe(unsafecast.Int32ToUint32(keys), values)
+}
+
+type Float32Table struct{ table32 }
+
+func NewFloat32Table(cap int, maxLoad float64) *Float32Table {
+	return &Float32Table{makeTable32(cap, maxLoad)}
+}
+
+func (t *Float32Table) Reset() { t.reset() }
+
+func (t *Float32Table) Len() int { return t.len }
+
+func (t *Float32Table) Cap() int { return t.cap }
+
+func (t *Float32Table) Probe(keys []float32, values []int32) int {
+	return t.probe(unsafecast.Float32ToUint32(keys), values)
+}
+
+type Uint32Table struct{ table32 }
+
+func NewUint32Table(cap int, maxLoad float64) *Uint32Table {
+	return &Uint32Table{makeTable32(cap, maxLoad)}
+}
+
+func (t *Uint32Table) Reset() { t.reset() }
+
+func (t *Uint32Table) Len() int { return t.len }
+
+func (t *Uint32Table) Cap() int { return t.cap }
+
+func (t *Uint32Table) Probe(keys []uint32, values []int32) int {
+	return t.probe(keys, values)
+}
+
+// table32 is the generic implementation of probing tables for 32 bit types.
+//
+// The table uses the following memory layout:
+//
+//		[bitmap][keys][values]
+//
+// - The bitmap is used to determine which slots of the table are occupied,
+//   when a bit is set to 1, the key and value at the corresponding index in
+//   the arrays of keys and values have been assigned.
+//
+// - The array of keys is written right after the bitmap. All keys are 32 bits
+//   values.
+//
+// - Values are written after the keys, with each slot of the keys array mapping
+//   to a value at the same slot in the values array.
+//
+// This memory layout is a combination between simplicity and performance; the
+// bitmap is required in order to differentiate between key slots with the zero
+// value, and key slots that have not yet been assigned.
```

Comment:
An alternative would to reserve the zero valued hash? You could update the hashing routines to rehash with a different (but derived) seed if the hash result is zero. A zero should be exceedingly rare, and a branch not taken in the hashing routines is essentially free. This would let you avoid the bitmaps entirely.

Reply:
I'll try it out.

Reply:
I started an experiment in https://github.com/segmentio/parquet-go/tree/hashprobe-no-zero-hash tho it appears that testing if the hash value is zero incurs a ~10% drop in throughput on dictionary operations:
```
name                     old time/op  new time/op  delta
Dictionary/Insert/INT32  2.55µs ± 0%  2.83µs ± 0%  +11.31%  (p=0.000 n=9+10)

name                     old value/s  new value/s  delta
Dictionary/Insert/INT32    393M ± 0%    353M ± 0%  -10.16%  (p=0.000 n=9+10)
```
It's possible that there is a balance between reducing the number of instructions in the probing routines which results in better throughput overall, the extra instructions on the hashing code paths seem expensive tho.

Reply:
I don't know if the branch is the problem, it might just be the extra instructions that add the cost.

AES hash for example does 3 `AESENC` instructions per loop which are 5 cycles each (15 cycles per loop plus a few for the loop increment and test).

To do a `TEST` or `CMP` we need to extract the value from a vector to a general purpose register, overall it may just cost too many CPU cycles in the hashing function to be worth removing cycles in the probing function.

I'm also a bit concerned about modifying the hashing algorithm; if we diverge, we cannot test against the Go runtime functions to ensure that our implementation is correct, and we might degrade the quality of the hash functions in ways that are hard to predict.

Reply:
There is a VPTEST instruction you can use to avoid the vector=>GP reg transfer cost, but yeah it might just be a case of too many extra instructions on the critical path.